File: tools/toolchain.py
#!/usr/bin/env python3
"""Explicit online environment staging and read-only offline toolchain preflight."""
import argparse
import hashlib
import json
import os
from pathlib import Path
import platform
import re
import shutil
import subprocess
import sys
import tempfile

ROOT = Path(__file__).resolve().parents[1]


def digest(path):
    with path.open("rb") as stream:
        return hashlib.file_digest(stream, "sha256").hexdigest()


def packages(kind):
    return json.loads((ROOT / f"config/{kind}-linux-64.provenance.json").read_text())["packages"]


def check_records(prefix, expected):
    actual = [json.loads(p.read_text()) for p in (prefix / "conda-meta").glob("*.json")]
    fields = ("name", "version", "build", "sha256")
    def records(rows):
        result = {}
        for row in rows:
            if not all(row.get(key) for key in fields) or row["name"] in result:
                raise ValueError(f"{prefix}: incomplete or duplicate package record")
            result[row["name"]] = tuple(row[key] for key in fields)
        return result
    wanted, found = records(expected), records(actual)
    if wanted != found:
        changed = sorted(k for k in wanted.keys() | found.keys() if wanted.get(k) != found.get(k))
        raise ValueError(f"{prefix}: package lock mismatch: {', '.join(changed)}")


def check_lock(path, expected):
    lines = [line.strip() for line in path.read_text().splitlines()
             if line.strip() and not line.startswith("#")]
    wanted = [row["url"] + "#" + row["md5"] for row in expected]
    if not lines or lines[0] != "@EXPLICIT" or sorted(lines[1:]) != sorted(wanted):
        raise ValueError(f"{path}: explicit URLs/checksums do not match provenance")


def new_destination(path):
    path = Path(os.path.abspath(path))
    if path.exists() or path.is_symlink():
        raise ValueError(f"destination already exists: {path}")
    if not path.parent.is_dir():
        raise ValueError(f"destination parent must exist: {path.parent}")
    return path


def environment():
    env = dict(os.environ)
    # Keep hidden library paths and user R profiles out of the pinned preflight.
    for key in ("LD_LIBRARY_PATH", "R_HOME", "R_LIBS", "R_LIBS_USER", "R_LIBS_SITE"):
        env.pop(key, None)
    # An unset R_LIBS_USER enables R's default ~/R library; /dev/null is never
    # a library directory, while the selected R installation keeps .Library.
    env["R_LIBS_USER"] = os.devnull
    env["R_LIBS_SITE"] = os.devnull
    for key in ("OMP_NUM_THREADS", "OPENBLAS_NUM_THREADS", "MKL_NUM_THREADS", "VECLIB_MAXIMUM_THREADS"):
        env[key] = "1"
    return env


def stage(args):
    if platform.system() != "Linux" or platform.machine() not in ("x86_64", "amd64"):
        raise ValueError("locks support Linux x86_64 only")
    prefix = new_destination(args.prefix)
    lock = ROOT / f"config/{args.kind}-linux-64.explicit.txt"
    check_lock(lock, packages(args.kind))
    manager = shutil.which(args.mamba)
    if manager is None:
        raise ValueError("mamba executable not found")
    argv = [manager, "create", "--yes", "--prefix", str(prefix), "--file", str(lock)]
    print(json.dumps({"online_staging_argv": argv}), flush=True)
    # Conda prefix relocation prevents creating elsewhere then renaming. An
    # interrupted/failed installation remains for inspection; never call it ready.
    subprocess.run(argv, check=True, env=environment())
    check_records(prefix, packages(args.kind))
    print(f"Staged and package-verified: {prefix}")


def preflight(args):
    alignment, rprefix = args.alignment.resolve(), args.r_prefix.resolve()
    destination = new_destination(args.out)
    for kind, prefix in (("alignment", alignment), ("p0", rprefix)):
        check_lock(ROOT / f"config/{kind}-linux-64.explicit.txt", packages(kind))
        check_records(prefix, packages(kind))
    manifest = ROOT / "vendor/toolchain/manifest.json"
    source_info = json.loads(manifest.read_text())
    for item in source_info["retained_files"]:
        path = ROOT / item["path"]
        if digest(path) != item["sha256"]:
            raise ValueError(f"retained recipe/license hash mismatch: {path}")
    env = environment()
    env["PATH"] = str(alignment / "bin") + os.pathsep + env.get("PATH", "")
    versions = [("STAR", "--version", r"(?m)^2\.7\.10b\s*$"),
                ("hisat2", "--version", r"version 2\.2\.1(?:\s|$)"),
                ("hisat2-build", "--version", r"version 2\.2\.1(?:\s|$)"),
                ("samtools", "--version", r"(?m)^samtools 1\.17\s*$"),
                ("featureCounts", "-v", r"featureCounts v2\.0\.6(?:\s|$)")]
    # Only the requested new artifact directory is modified by preflight.
    scratch = Path(tempfile.mkdtemp(prefix=".preflight-", dir=destination.parent))
    try:
        observed = []
        for name, flag, pattern in versions:
            binary = alignment / "bin" / name
            result = subprocess.run([str(binary), flag], capture_output=True, text=True,
                                    errors="replace", timeout=30, env=env)
            output = result.stdout + result.stderr
            (scratch / f"{name}.version.txt").write_text(output)
            if result.returncode or not re.search(pattern, output):
                raise ValueError(f"{name}: pinned version probe failed: {output}")
            observed.append({"name": name, "executable": str(binary), "sha256": digest(binary),
                             "version_output": output})
        with (scratch / "R.log").open("w") as log:
            subprocess.run([str(rprefix / "bin/Rscript"), "--vanilla",
                            str(ROOT / "tools/preflight_r.R"), str(scratch)],
                           stdout=log, stderr=subprocess.STDOUT, check=True, timeout=180, env=env)
        report = {"platform": platform.platform(), "alignment_prefix": str(alignment),
                  "r_prefix": str(rprefix), "tools": observed,
                  "tool_manifest_sha256": digest(manifest),
                  "locks": {kind: digest(ROOT / f"config/{kind}-linux-64.explicit.txt")
                            for kind in ("alignment", "p0")},
                  "scope": "Installed package records, versions, R computation and PNG; no installation or fetch."}
        (scratch / "verified.json").write_text(json.dumps(report, indent=2) + "\n")
        if destination.exists() or destination.is_symlink():
            raise ValueError("output appeared during preflight")
        scratch.rename(destination)
    except BaseException:
        print(f"Failed preflight diagnostics retained at {scratch}", file=sys.stderr)
        raise
    print(f"PASS: offline toolchain and R preflight: {destination}")


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    commands = parser.add_subparsers(dest="command", required=True)
    install = commands.add_parser("stage", help="explicit ONLINE installation; never invoked by compute")
    install.add_argument("kind", choices=("alignment", "p0"))
    install.add_argument("--prefix", type=Path, required=True)
    install.add_argument("--mamba", default="mamba")
    check = commands.add_parser("preflight", help="OFFLINE package/version/R/PNG checks; no install")
    check.add_argument("--alignment", type=Path, default=ROOT / ".deps/alignment")
    check.add_argument("--r-prefix", type=Path, default=ROOT / ".deps/p0")
    check.add_argument("--out", type=Path, required=True)
    args = parser.parse_args()
    try:
        (stage if args.command == "stage" else preflight)(args)
    except (OSError, ValueError, subprocess.SubprocessError) as error:
        parser.exit(1, f"ERROR: {error}\n")


if __name__ == "__main__":
    main()

File: tools/toolchain.R
#!/usr/bin/env Rscript
source(file.path(dirname(sub('^--file=','',grep('^--file=',commandArgs(),value=TRUE)[1])),if(basename(dirname(sub('^--file=','',grep('^--file=',commandArgs(),value=TRUE)[1])))=='integration')'../../tools/lib_helpers.R' else 'lib_helpers.R'))
packages <- function(kind) read_json(file.path(ROOT,paste0('config/',kind,'-linux-64.provenance.json')))$packages
check_records <- function(prefix, expected) {
 records<-function(rows) {
  result<-list(); fields<-c('name','version','build','sha256')
  for(row in rows) { assert(all(vapply(fields,function(k)!is.null(row[[k]])&&nzchar(row[[k]]),TRUE))&&!row$name %in% names(result),'incomplete or duplicate package record'); result[[row$name]]<-row[fields] }
  result[sort(names(result))]
 }
 actual<-lapply(list.files(file.path(prefix,'conda-meta'),'[.]json$',full.names=TRUE),read_json)
 # Explicit MD5 installs may omit SHA256 from conda-meta and record a legacy
 # tar.bz2 cache suffix even for .conda archives. Verify the actual retained
 # archive against the pinned SHA256 instead of weakening the identity check.
 for(i in seq_along(actual)) if(is.null(actual[[i]]$sha256) || !nzchar(actual[[i]]$sha256)) {
  row<-actual[[i]]; match<-Filter(function(w)identical(w$name,row$name),expected)
  assert(length(match)==1L,'unexpected installed package');wanted<-match[[1]]
  assert(identical(row$url,wanted$url)&&identical(row$md5,wanted$md5),'installed package URL/MD5 mismatch')
  cached<-row$package_tarball_full_path
  assert(!is.null(cached)&&nzchar(cached),'missing retained package archive location')
  candidates<-unique(c(cached,file.path(dirname(cached),basename(wanted$url))))
  candidates<-candidates[file.exists(candidates)&!dir.exists(candidates)]
  assert(length(candidates)>0L,paste('SHA256 absent; retain pinned package archive for',row$name))
  actual[[i]]$sha256<-sha256(candidates[1])
 }
 assert(identical(records(expected),records(actual)),paste(prefix,'package lock mismatch'))
}
check_lock <- function(path, expected) {
 lines<-trimws(readLines(path)); lines<-lines[nzchar(lines)&!startsWith(lines,'#')]
 wanted<-vapply(expected,function(row)paste0(row$url,'#',row$md5),'')
 assert(length(lines)>0L && lines[1]=='@EXPLICIT' && identical(sort(lines[-1]),sort(wanted)),paste(path,'explicit URLs/checksums do not match provenance'))
}
stage_toolchain <- function(kind,prefix,mamba='mamba') {
 assert(Sys.info()[['sysname']]=='Linux' && Sys.info()[['machine']] %in% c('x86_64','amd64'),'locks support Linux x86_64 only')
 prefix<-new_destination(prefix); lock<-file.path(ROOT,paste0('config/',kind,'-linux-64.explicit.txt')); check_lock(lock,packages(kind))
 manager<-Sys.which(mamba); assert(nzchar(manager),'mamba executable not found')
 cat(jsonlite::toJSON(list(online_staging_argv=c(manager,'create','--yes','--prefix',prefix,'--file',lock)),auto_unbox=TRUE),'\n')
 with_environment(clean_environment(),run(c(manager,'create','--yes','--prefix',prefix,'--file',lock)))
 check_records(prefix,packages(kind)); cat('Staged and package-verified:',prefix,'\n')
}
preflight_toolchain <- function(alignment,rprefix,destination,r_kind='runtime-r') {
 alignment<-absolute(alignment); rprefix<-absolute(rprefix); destination<-new_destination(destination)
 check_lock(file.path(ROOT,paste0('config/',r_kind,'-linux-64.explicit.txt')),packages(r_kind)); check_records(rprefix,packages(r_kind))
 if(dir.exists(file.path(alignment,'conda-meta'))) {
  check_lock(file.path(ROOT,'config/alignment-linux-64.explicit.txt'),packages('alignment')); check_records(alignment,packages('alignment'))
 } else {
  # Runtime bundle schema is validated independently of historical conda records.
  assert(file.exists(file.path(alignment,'COMPLETE')),'incomplete native runtime bundle')
  bundle<-table_read(file.path(alignment,'provenance/files.tsv')); assert(nrow(bundle)>0L,'empty native runtime manifest')
  for(item in rows_list(bundle)) { assert(!grepl('(^/|(^|/)[.][.](/|$))',item$path),'unsafe runtime manifest path'); assert(sha256(file.path(alignment,item$path))==item$sha256,'runtime file hash mismatch') }
  local({ old<-getwd(); on.exit(setwd(old)); setwd(alignment); run(c('sha256sum','-c','SHA256SUMS')) })
  assert(!any(grepl('(^|/)(python[^/]*|snakemake)$',list.files(alignment,recursive=TRUE,all.files=TRUE))),'Python/Snakemake found in runtime bundle')
 }
 manifest<-file.path(ROOT,'vendor/toolchain/manifest.json'); evidence<-read_json(manifest)
 for(item in evidence$retained_files) assert(sha256(file.path(ROOT,item$path))==item$sha256,paste('retained recipe/license hash mismatch:',item$path))
 scratch<-new_directory('.preflight-',dirname(destination)); okay<-FALSE
 on.exit(if(!okay)cat('Failed preflight diagnostics retained at',scratch,'\n',file=stderr()))
 probes<-list(STAR=c('--version','^2[.]7[.]10b$'),'hisat2-align-s'=c('--version','(^|/)hisat2-align-s version 2[.]2[.]1$'),'hisat2-build-s'=c('--version','(^|/)hisat2-build-s version 2[.]2[.]1$'),samtools=c('--version','^samtools 1[.]17$'),featureCounts=c('-v','^featureCounts v2[.]0[.]6$'))
 observed<-with_environment(c(clean_environment(),list(PATH=paste(file.path(alignment,'bin'),Sys.getenv('PATH'),sep=':'))),{
  values<-list()
  for(name in names(probes)) { binary<-file.path(alignment,'bin',name); output<-run(c(binary,probes[[name]][1]),timeout=30)$output
   writeLines(output,file.path(scratch,paste0(name,'.version.txt'))); assert(any(grepl(probes[[name]][2],output)),paste(name,'pinned version probe failed'))
   values[[length(values)+1L]]<-list(name=name,executable=binary,sha256=sha256(binary),version_output=paste(output,collapse='\n')) }
  run(c(file.path(rprefix,'bin/Rscript'),'--vanilla',file.path(ROOT,'tools/preflight_r.R'),scratch),log=file.path(scratch,'R.log'),timeout=180); values
 })
 write_json(list(platform=as.list(Sys.info()),alignment_prefix=alignment,r_prefix=rprefix,tools=observed,tool_manifest_sha256=sha256(manifest),r_lock_sha256=sha256(file.path(ROOT,paste0('config/',r_kind,'-linux-64.explicit.txt'))),scope='Installed records, native runtime hashes, exact versions, R computation and PNG; no fetch or install.'),file.path(scratch,'verified.json'))
 assert(!exists_any(destination),'output appeared during preflight'); assert(file.rename(scratch,destination),'cannot publish preflight'); okay<-TRUE
 cat('PASS offline toolchain and R preflight:',destination,'\n')
}
main <- function() {
 argv<-commandArgs(TRUE); assert(length(argv)>0L,'Usage: toolchain.R stage KIND --prefix PATH | verify --kind KIND --prefix PATH | preflight --out PATH')
 command<-argv[1]; argv<-argv[-1]
 if(command=='stage') { assert(length(argv)>0L,'missing kind'); kind<-argv[1]; args<-parse_cli(list(prefix=NULL,mamba='mamba'),required='prefix',args=argv[-1]); stage_toolchain(kind,args$prefix,args$mamba) }
 else if(command=='verify') { args<-parse_cli(list(kind=NULL,prefix=NULL),required=c('kind','prefix'),args=argv); check_lock(file.path(ROOT,paste0('config/',args$kind,'-linux-64.explicit.txt')),packages(args$kind)); check_records(args$prefix,packages(args$kind)); cat('PASS package lock and installed records\n') }
 else if(command=='preflight') { args<-parse_cli(list(alignment=file.path(ROOT,'.deps/runtime-tools'),'r-prefix'=file.path(ROOT,'.deps/runtime-r'),'r-kind'='runtime-r',out=NULL),required='out',args=argv); preflight_toolchain(args$alignment,args[['r-prefix']],args$out,args[['r-kind']]) }
 else stop('unknown command')
}
if(sys.nframe()==0L)main()
